R/rules.R

```diff
@@ -125,3 +125,101 @@ as.list.rule <- function(x, ...) {
 
   r_list
 }
+
+#' Combine Two Rules
+#'
+#' @param x (`rule`) to modify.
+#' @param y (`rule`) rule whose mapping will take precedence over the ones described in `x`.
+#' @param ... not used.
+#'
+#' @note The order of the mappings in the resulting rule corresponds to the order of the mappings in `x` followed by the
+#'   mappings that are only present in `y`.
+#'
+#' @returns a `rule`.
+#' @export
+#' @examples
+#' r1 <- rule(
+#'   "first" = c("from ori rule", "FROM ORI RULE"),
+#'   "last" = c(NA, "last"),
+#'   .to_NA = "X",
+#'   .drop = TRUE
+#' )
+#' r2 <- rule(
+#'   "first" = c("F", "f"),
+#'   "second" = c("S", "s"),
+#'   "third" = c("T", "t"),
+#'   .to_NA = "something"
+#' )
+#' combine_rules(r1, r2)
+combine_rules <- function(x, y, ...) {
+  checkmate::assert_class(x, "rule", null.ok = TRUE)
+  checkmate::assert_class(y, "rule", null.ok = TRUE)
+
+  if (is.null(x) && is.null(y)) {
+    rlang::abort("Both rules are NULL.")
+  }
+
+  # If one of the rules is NULL, return the other (via empty list).
+  x <- as.list(x)
+  y <- as.list(y)
+  names_y <- names(y)
+  names_x <- names(x)
+  names_x_diff <- setdiff(names_x, names_y)
+
+  x <- x[names_x_diff]
+  r <- c(y, x)
+
+  # Reorder to follow original order
+  names_y_diff <- setdiff(names_y, names_x)
+  names_order <- c(names_x, names_y_diff)
+  r <- r[names_order]
```

Comment:
current structure seems to be same as `modifyList(x, y)` or `x[names(y)] <- y` (which keeps the order of x)